File: commands/probability/howmanyruns.ts
import { ChatInputCommandInteraction, EmbedBuilder, SlashCommandSubcommandBuilder } from 'discord.js';
import Bot from '../../bot/bot';
import SlashCommand from '../../bot/slashcommand';

type Stringifiable = {
  toString(): string;
}

const formatWithSpace = (n: number | string) => {
  const parts = n.toString().split('.');
  parts[0] = parts[0].replace(/\B(?=(\d{3})+(?!\d))/g, ' ');
  return parts.join('.');
};
const transpose = <T>(matrix: Array<Array<T>>) => matrix[0].map((_, i) => matrix.map(row => row[i]));
const codeblock = (s: string) => `\`\`\`${s}\`\`\``;

const longestString = <T extends Stringifiable>(arr: Array<T>) => Math.max(...arr.map(e => e.toString().length));
const padStrings = <T extends Stringifiable>(arr: Array<T>, minLength?: number) => arr.map(s => s.toString().padEnd(Math.max(minLength ?? 0, longestString(arr) + 4)));

export const verifyProbability = (input: string) => {
  switch (true) {
    case /^0\.\d*[1-9]$/.test(input): {
      // Number is on form 0.123
      return parseFloat(input);
    }
    case /^[1-9]\d*\/[1-9]\d*$/.test(input): {
      // Number is on form x/y
      const split = input.split('/');
      const result = parseInt(split[0]) / parseInt(split[1]);
      // Require: result ∈ (0,1)
      return result > 0 && result < 1 ? result : -1;
    }
    default: {
      return -1;
    }
  }
};

export default class HowManyRuns extends SlashCommand {
  constructor(client: Bot) {
    const data = new SlashCommandSubcommandBuilder()
      .setName('howmanyruns')
      .setDescription('Calculate number of runs required to obtain an item with a given drop chance')
      .addStringOption(option => option
        .setName('probability')
        .setDescription('A decimal number or fraction between 0 and 1 (not inclusive)')
        .setRequired(true)
      )
      .addNumberOption(option => option
        .setName('items-per-run')
        .setDescription('The number of items you get in a single run')
        .setMinValue(1)
      )
      .addNumberOption(option => option
        .setName('time-per-run')
        .setDescription('The time (in minutes) to complete a single run')
        .setMinValue(0.01)
      );
    super(data, client, { belongsTo: 'probability' });
  }

  async execute(interaction: ChatInputCommandInteraction): Promise<void> {
    const probabilityInput = interaction.options.getString('probability')!;
    const itemsPerRun = interaction.options.getNumber('items-per-run');
    const timePerRun = interaction.options.getNumber('time-per-run');

    const probability = verifyProbability(probabilityInput);

    if (probability === -1) {
      const embed = new EmbedBuilder()
        .setColor(this.client.config.colors.RED)
        .setTitle('Invalid Format')
        .setDescription('The argument \`probability\` must be a decimal number or a fraction, and be between 0 and 1.');
      interaction.reply({ embeds: [embed], ephemeral: true });
      return;
    }

    const probabilities = [10, 25, 50, 75, 90, 95, 99] as const;
    
    /**
     * Probability p of getting at least 1 success in n runs \
     * p = 1 - (1 - probabilityOfSuccess) ^ n \
     * n = ln(1 - p) / ln(1 - probabilityOfSuccess)
     * @returns n
     */
    const attemptsRequired = (p: number) => Math.round(Math.log(1 - p) / Math.log(1 - probability));
    
    const itemsRequired = probabilities.map(p => attemptsRequired(p / 100));

    const data: Array<Array<string | number>> = [
      padStrings(['Prob', ...probabilities.map(n => n + '%')]),
      padStrings(['Items', ...itemsRequired.map(n => formatWithSpace(n))], 6)
    ];

    let description = 'Items: The total amount of items to have an X% chance of at least 1 rare drop';

    if (itemsPerRun) {
      description += `\nRuns: The number of runs required, assuming **${itemsPerRun}** items per run`;
      const runsRequired = itemsRequired.map(itemCount => formatWithSpace(Math.round(itemCount / itemsPerRun)));
      data.push(padStrings(['Runs', ...runsRequired], 6));
    }
    if (timePerRun) {
      description += `\nHours: The number of hours spent farming, assuming one run takes **${timePerRun}** minutes`;
      const hoursRequired = itemsRequired.map(itemCount => formatWithSpace((itemCount * timePerRun / (60 * (itemsPerRun ?? 1))).toFixed(1)));
      data.push(padStrings(['Hours', ...hoursRequired], 6));
    }
    
    const prepared = transpose(data).map(row => row.join(''));
    prepared.splice(1, 0, ''); // Insert a '' at index 1
    const content = codeblock(prepared.join('\n'));

    const embed = new EmbedBuilder()
      .setColor(this.client.config.colors.BLUE)
      .setTitle('Drop Chance Analysis')
      .setDescription(description)
      .addFields({
        name: `For an item with drop chance **${probabilityInput}**, here are the relevant rates.`,
        value: content
      });

    interaction.reply({ embeds: [embed] });
  }
}

File: src/commands/probability/howmanyruns.rs
use regex::Regex;
use serenity::all::{CommandInteraction, CommandOptionType};
use serenity::async_trait;
use serenity::builder::{CreateCommandOption, CreateEmbed};
use serenity::client::Context;

use crate::color::Colors;
use crate::commands::SlashSubCommand;
use crate::database::Database;
use crate::interaction::{BetterResponse, InteractionCustomGet};
use crate::Result;

#[derive(Default)]
pub struct HowManyRuns;

#[async_trait]
impl SlashSubCommand for HowManyRuns {
  fn register(&self) -> CreateCommandOption {
    CreateCommandOption::new(
      CommandOptionType::SubCommand,
      "howmanyruns",
      "Calculate number of runs required to obtain an item with a given drop chance"
    )
    .add_sub_option(
      CreateCommandOption::new(
        CommandOptionType::String,
        "probability",
        "A decimal number or fraction between 0 and 1 (not inclusive)"
      )
      .required(true)
    )
    .add_sub_option(
      CreateCommandOption::new(
        CommandOptionType::Integer,
        "items-per-run",
        "The number of items you get in a single run/attempt"
      )
      .min_int_value(1)
    )
    .add_sub_option(
      CreateCommandOption::new(
        CommandOptionType::Number,
        "minutes-per-run",
        "The number of minutes to complete a single run"
      )
      .min_number_value(0.01)
    )
  }

  async fn execute(&self, ctx: &Context, interaction: &CommandInteraction, _: &Database) -> Result<()> {
    let probability_input = interaction.get_string("probability").unwrap().replace(' ', "");
    let items_per_run = interaction.get_integer("items-per-run");
    let time_per_run = interaction.get_number("minutes-per-run");

    let probability = match verify_probability(probability_input.as_str()) {
      Some(parsed) => parsed,
      None => {
        let embed = CreateEmbed::new()
          .color(Colors::Red)
          .title("Invalid format")
          .description("The argument `probability` must be a decimal number or a fraction, and be between 0 and 1.");

        interaction.reply_embed_ephemeral(ctx, embed).await?;
        return Ok(());
      }
    };

    // TODO: Let user specify own set of sample percentages (basically customize this list)
    let probabilities = [10, 25, 50, 75, 90, 95, 99];

    //
    // Probability p of getting at least 1 success in n runs
    // p = 1 - (1 - probabilityOfSuccess) ^ n
    //   =>
    // n = ln(1 - p) / ln(1 - probabilityOfSuccess)
    //
    let attempts_required = |p: f32| ((1_f32 - p).ln() / (1_f32 - probability).ln()).round() as i32;

    let items_required = probabilities
      .iter()
      .map(|&p| attempts_required(p as f32 / 100_f32))
      .collect::<Vec<_>>();

    let mut probabilities_display = probabilities.iter().map(|p| format!("{p}%")).collect::<Vec<_>>();
    probabilities_display.insert(0, "Prob".to_string());

    let mut items_display = items_required
      .iter()
      .map(|&n| format_float(n as f32, 0))
      .collect::<Vec<_>>();
    items_display.insert(0, "Items".to_string());

    let mut data = vec![pad_strings(&probabilities_display), pad_strings(&items_display)];

    let mut description =
      vec!["**Items**: The total amount of items to have an X% chance of at least 1 rare drop".to_string()];

    if let Some(items) = items_per_run {
      description.push(format!(
        "**Runs**: The amount of runs required, assuming **{items}** items per run"
      ));
      let mut runs_required = items_required
        .iter()
        .map(|&n| format_float(n as f32 / items as f32, 0))
        .collect::<Vec<_>>();
      runs_required.insert(0, "Runs".to_string());
      data.push(pad_strings(&runs_required));
    }

    if let Some(time) = time_per_run {
      description.push(format!(
        "**Hours**: The time spent farming, assuming one run takes **{}** minutes",
        time
      ));
      let mut time_required = items_required
        .iter()
        .map(|&n| {
          let time = n as f32 * time as f32 / (60_f32 * items_per_run.unwrap_or(1) as f32);
          format_float(time, 1)
        })
        .collect::<Vec<_>>();
      time_required.insert(0, "Hours".to_string());
      data.push(pad_strings(&time_required));
    }

    let mut prepared = transpose(data).iter().map(|row| row.join("   ")).collect::<Vec<_>>();
    prepared.insert(1, "".to_string()); // Spacer
    let content = format!("```{}```", prepared.join("\n"));

    let embed = CreateEmbed::new()
      .color(Colors::Blue)
      .title("Drop Chance Analysis")
      .description(description.join("\n"))
      .field(
        format!("For an item with a drop chance of **{probability_input}**, here are the relevant rates"),
        content,
        false
      );

    interaction.reply_embed(ctx, embed).await?;
    Ok(())
  }
}

pub fn verify_probability(input: &str) -> Option<f32> {
  let decimal = Regex::new(r"^0\.\d*[1-9]$").unwrap(); // On form 0.123
  let fraction = Regex::new(r"^[1-9]\d*\/[1-9]\d*$").unwrap(); // On form x/y

  if decimal.is_match(input) {
    Some(input.parse::<f32>().unwrap())
  } else if fraction.is_match(input) {
    let split = input.split('/').collect::<Vec<&str>>();
    let result = split[0].parse::<f32>().unwrap() / split[1].parse::<f32>().unwrap();

    // Require result ∈ (0,1)
    if result > 0.0 && result < 1.0 {
      Some(result)
    } else {
      None
    }
  } else {
    None
  }
}

fn pad_strings<T: ToString>(array: &[T]) -> Vec<String> {
  let length = longest_string(array);

  array
    .iter()
    .map(|e| format!("{: >1$}", e.to_string(), length))
    .collect::<Vec<_>>()
}

fn longest_string<T: ToString>(array: &[T]) -> usize {
  array.iter().map(|e| e.to_string().len()).max().unwrap()
}

fn transpose<T: Clone>(matrix: Vec<Vec<T>>) -> Vec<Vec<T>> {
  (0..matrix[0].len())
    .map(|i| matrix.iter().map(|inner| inner[i].clone()).collect::<Vec<T>>())
    .collect()
}

/// Format a floating point number to a specific precision, spacing out every 3rd digit
///
/// Examples: (`31415.9265`, `2`) -> `"31 415.93"`
fn format_float(input: f32, decimal_count: usize) -> String {
  let scaled = input * 10_i32.pow(decimal_count as u32) as f32;
  let literal = scaled.round().to_string();
  let reversed = literal.chars().rev().collect::<String>();
  let mut data = format!("{:0<1$}", reversed, decimal_count + 1);

  let mut index = if decimal_count > 0 {
    data.insert(decimal_count, '.');
    decimal_count + 4
  } else {
    decimal_count + 3
  };

  while index < data.len() {
    data.insert(index, ' ');
    index += 4;
  }

  data.chars().rev().collect::<String>()
}
